### ISSUE TITLE
XMPP SASL DIGEST-MD5 authentication breaks with jabberd 2.1.4

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2606]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#2606|
|Type|enhancement|
|Created|2007-04-25 13:46:50Z|

The decoding of the challenge is done too naively, so that it breaks on multiple values for the `qop` and `cipher` directives.

<details><summary>Searchable metadata</summary>

```
trac-id__2606 2606
type__enhancement enhancement
reporter__ralphm ralphm
priority__highest highest
milestone__Words_0_6 Words-0.6
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1177508810000000 1177508810000000
changetime__1177515403000000 1177515403000000
version__None None
owner__ 
cc__ralphm cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_1"></a>|@ralphm removed owner|
|-|-|

While waiting for [#2015](https://github.com/twisted/twisted/issues/2015) to be completed, I took the liberty of taking its parsing code to fix this problem.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to @ralphm|
|-|-|

 * test!__parse should be renamed to test_parse
 * I think the comment in _parse should say 'because' instead of 'before'
 * the code at the end of _parse would be better written this way IMO:
```
for param in ['qop', 'cipher']:
    if param in paramDict:
        paramDict[param] = paramDict[param].split(',')
```
 * Please upgrade the copyright of sasl_mechanisms to 2007 and add one to test_jabbersaslmechanisms
 * If you can, remove the call to has_key in sasl_mechanisms

Outside of this, that seems reasonable. I'm not sure how it's useful since _gen_response doesn't use these values anyway, but I guess it's in the todo list.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_3"></a>|@pitrou commented|
|-|-|

The patch looks fine to me (although of course I'd prefer [#2015](https://github.com/twisted/twisted/issues/2015) to be reviewed ;-)).

therve: it's useful because otherwise parsing of other parameters fails (the comma in the qop value is mistaken as a parameter separator).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm commented|
|-|-|

Thanks for the comments.

I suppose it should really do something with the values down the road.

Committed the changes. Merging.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_5"></a>|@ralphm set status to `closed`|
|-|-|

(In [20070]) Unbreak XMPP SASL DIGEST-MD5 challenge parsing with multi-valued directives.

Author: ralphm
Reviewer: therve (antoine)
Fixes [#2606](https://github.com/twisted/twisted/issues/2606).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
